### PR TITLE
Filter Page Editor variable popover based on current variable name

### PR DESCRIPTION
**Pull request by twschiller**
## What does this PR do?

- Implements filtering and other improvements for the variable popover
- Fixes a bug where variable input mode switched to text when you start typing a variable name, due to invalid intermediate states

## Reviewer Notes

- You must enable the popover in the Skunkworks to use. (Variable popover is still a skunkworks feature)
- `TextWidget` contains fixes for triggering the popover

## Other Variable Popover Improvements
- [x] Filter-top level
- [x] Filter sub-levels
- [x] Only auto-expand current level
- [x] Support arrays
- [x] Show empty state
- [x] Highlight row on hover
- [x] Make the whole row clickable
- [x] Hide integration variables 
- [x] If no match at top level, return all variables at top level so user can see which variables are valid
- [x] Fix sort: still wrong for doubly-nested properties? - it was working, but confusing since `attrs` and `data` were empty objects
- [x] Don't show variable/template warnings while popover is active
- [x] Use the keyboard up/down arrows to navigate the popover modal
- [x] Support tab to choose the highlighted entry

## Demo

- https://www.loom.com/share/e6f65195c88c4531a627ab02cae91bc7?sid=94303cf1-f9ed-49db-a2a4-84d1f073d036

## Future Work

- Consider the keyboard cursor position when filtering. (Only consider up to the current character, that allows the users to expand the search by using the left arrow)
- Show breadcrumb vs. nested parents in order to preserve more space? In IntelliJ autocomplete, the don't repeat the parent props in the auto-complete
- Improve style/make more compact

## Checklist

- [x] Add tests
- [x] Designate a primary reviewer: @grahamlangford

### CODE REVIEW

File: src/components/fields/schemaFields/widgets/TextWidget.tsx [hunk 1]
```diff
@@ -56,27 +57,51 @@ function schemaSupportsTemplates(schema: Schema): boolean {
...
+ * Return true if the value is a valid variable expression
+ */
 export function isVarValue(value: string): boolean {
   return objectPathRegex.test(value);
 }
 
+/**
+ * Returns true if the value is a valid variable expression or var-like expression while the use is typing
+ */
+export function isVarLike(value: string): boolean {
```
Comment (twschiller) [inline src/components/fields/schemaFields/widgets/TextWidget.tsx:87]: Fixes intermediary state to capturing likely user intent

File: src/components/fields/schemaFields/widgets/varPopup/VariablesTree.test.tsx [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+    const wrapper = render(
+      <VariablesTree
+        vars={inputMap}
+        onVarSelect={noop}
+        likelyVariable="@input.foo."
+      />
+    );
+
+    const elements = wrapper.getAllByRole("button");
```
Comment (BLoe) [inline src/components/fields/schemaFields/widgets/varPopup/VariablesTree.test.tsx:75]: Can we use `screen` here?

File: src/components/fields/schemaFields/widgets/varPopup/VariablesTree.tsx [hunk 3]
```diff
@@ -108,53 +36,80 @@ const defaultItemString: GetItemString = (type, data, itemType, itemString) => (
 
 const getItemString: GetItemString = (type, data, itemType, itemString) => {
   if (type === "Object" && isEmpty(data)) {
-    return <span>not set</span>;
+    // Corresponds to entries where not trace value exists
```
Comment (BLoe) [inline src/components/fields/schemaFields/widgets/varPopup/VariablesTree.tsx:39]:
```suggestion
    // Corresponds to entries where trace value does not exist
```

**Review by grahamlangford — APPROVED** [review]

**Review by BLoe — APPROVED** [review]

File: src/components/fields/schemaFields/widgets/varPopup/VarMenu.tsx [hunk 4]
```diff
@@ -58,19 +90,18 @@ const VarMenu: React.FunctionComponent<VarMenuProps> = ({
...
-  const pipelineMap = useSelector(selectPipelineMap) ?? {};
-  const { allBlocks } = useAllBlocks();
+  useEffect(() => {
+    console.debug("showVariablePopover");
+    dispatch(editorActions.showVariablePopover(null));
 
-  const knownVars = useSelector(selectKnownVarsForActiveNode);
-  const trace = useSelector(selectActiveNodeTrace);
+    return () => {
+      console.debug("hideVariablePopover");
```
[the remark this answers is not in the document]
Comment (twschiller) [reply src/components/fields/schemaFields/widgets/varPopup/VarMenu.tsx:98]: I'll remove them. They were for mount and unmount

File: src/components/fields/schemaFields/widgets/varPopup/VarMenu.tsx [hunk 5]
```diff
@@ -58,19 +90,18 @@ const VarMenu: React.FunctionComponent<VarMenuProps> = ({
     };
   }, [onClose]);
 
-  const activeElement = useSelector(selectActiveElement);
-  const pipelineMap = useSelector(selectPipelineMap) ?? {};
-  const { allBlocks } = useAllBlocks();
+  useEffect(() => {
+    console.debug("showVariablePopover");
+    dispatch(editorActions.showVariablePopover(null));
```
[the remark this answers is not in the document]
Comment (twschiller) [reply src/components/fields/schemaFields/widgets/varPopup/VarMenu.tsx:95]: Hmm, I thought typescript was yelling at me

**Pull request merged**